{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "45c92254-ec10-4c5a-a1dd-5fe653d9a09c",
   "metadata": {},
   "source": [
    "# Code Quality Checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "48631d77-6dce-456b-a58d-e663a8eb1fd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../')\n",
    "try:\n",
    "    from premier_league import constants\n",
    "    from premier_league import code_quality_functions\n",
    "except ModuleNotFoundError:\n",
    "    import constants\n",
    "    import code_quality_functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "37a098e9-8ebd-44aa-96fa-28fc797f01aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/david@inawisdom.com/Documents/Training/premier_league/notebooks'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "852039e9-89fb-4908-8514-30e349a869cc",
   "metadata": {},
   "source": [
    "## Unit Tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "370a3694-60e4-4c2f-859f-8c5d10014840",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "============================= test session starts ==============================\n",
      "platform darwin -- Python 3.10.1, pytest-7.4.3, pluggy-1.3.0 -- /Users/david@inawisdom.com/Documents/Training/premier_league/env_premier_league/bin/python\n",
      "metadata: {'Python': '3.10.1', 'Platform': 'macOS-14.1-arm64-arm-64bit', 'Packages': {'pytest': '7.4.3', 'pluggy': '1.3.0'}, 'Plugins': {'html': '2.0.1', 'rerunfailures': '12.0', 'xdist': '3.4.0', 'cov': '4.1.0', 'seleniumbase': '4.21.4', 'metadata': '3.0.0', 'mock': '3.12.0', 'anyio': '3.7.1', 'ordering': '0.6'}}\n",
      "rootdir: /Users/david@inawisdom.com/Documents/Training/premier_league\n",
      "plugins: html-2.0.1, rerunfailures-12.0, xdist-3.4.0, cov-4.1.0, seleniumbase-4.21.4, metadata-3.0.0, mock-3.12.0, anyio-3.7.1, ordering-0.6\n",
      "collecting ... collected 4 items\n",
      "\n",
      "../tests/test_s3_helpers.py::test_grab_data_s3[some/file/path.csv-some-bucket-col1,col2\\n1,2\\n3,4\\n-expected_output0] PASSED [ 25%]\n",
      "../tests/test_s3_helpers.py::test_save_data_s3[data0-path/to/file.csv-test-bucket-csv-custom-profile-csv-None] PASSED [ 50%]\n",
      "../tests/test_s3_helpers.py::test_save_data_s3[data1-path/to/file.xlsx-test-bucket-excel-None-excel-None] PASSED [ 75%]\n",
      "../tests/test_s3_helpers.py::test_save_data_s3[data2-path/to/file.txt-test-bucket-txt-None-None-ValueError] FAILED [100%]\n",
      "\n",
      "=================================== FAILURES ===================================\n",
      "_ test_save_data_s3[data2-path/to/file.txt-test-bucket-txt-None-None-ValueError] _\n",
      "\n",
      "data =    a\n",
      "0  1\n",
      "1  2, file_path = 'path/to/file.txt', bucket = 'test-bucket'\n",
      "file_format = 'txt', profile_name = None\n",
      "\n",
      "    def save_data_s3(\n",
      "        data: pd.DataFrame,\n",
      "        file_path: str,\n",
      "        bucket: str = constants.S3_BUCKET,\n",
      "        file_format: str = \"csv\",\n",
      "        profile_name: Optional[str] = \"premier-league-app\",\n",
      "    ) -> None:\n",
      "        \"\"\"\n",
      "        Save a DataFrame to an S3 bucket.\n",
      "    \n",
      "        Parameters:\n",
      "        - data (pd.DataFrame): DataFrame to save.\n",
      "        - file_path (str): Path to the file within the S3 bucket.\n",
      "        - bucket (str): Name of the S3 bucket. Defaults to the value in\n",
      "        constants.S3_BUCKET.\n",
      "        - file_format (str): Format of the file to save ('csv', 'excel', etc.). Defaults\n",
      "        to \"csv\".\n",
      "        - profile_name (Optional[str]): AWS profile name. Defaults to \"premier-league-app\".\n",
      "    \n",
      "        Raises:\n",
      "        - NoCredentialsError: If AWS credentials are missing.\n",
      "        - PartialCredentialsError: If AWS credentials are incomplete.\n",
      "        \"\"\"\n",
      "        try:\n",
      "            if profile_name:\n",
      "                # Use the specified profile\n",
      "                session = boto3.Session(profile_name=profile_name)\n",
      "            else:\n",
      "                # Use environment variables or default profile\n",
      "                session = boto3.Session(\n",
      "                    aws_access_key_id=os.getenv(\"AWS_ACCESS_KEY_ID\"),\n",
      "                    aws_secret_access_key=os.getenv(\"AWS_SECRET_ACCESS_KEY\"),\n",
      "                    region_name=os.getenv(\"AWS_DEFAULT_REGION\"),\n",
      "                )\n",
      "            s3_resource = session.resource(\"s3\")\n",
      "            logger_config.logger.info(f\"Savng data to {bucket}/{file_path}\")\n",
      "            if file_format.lower() == \"csv\":\n",
      "                # Save DataFrame as a CSV\n",
      "                buffer = StringIO()\n",
      "                data.to_csv(buffer, index=False)\n",
      "                buffer.seek(0)\n",
      "                s3_resource.Bucket(bucket).put_object(Key=file_path, Body=buffer.getvalue())\n",
      "            elif file_format.lower() == \"excel\":\n",
      "                # Save DataFrame as an Excel file\n",
      "                buffer = BytesIO()\n",
      "                with pd.ExcelWriter(buffer, engine=\"xlsxwriter\") as writer:\n",
      "                    data.to_excel(writer, index=False)\n",
      "                buffer.seek(0)\n",
      "                s3_resource.Bucket(bucket).put_object(Key=file_path, Body=buffer.getvalue())\n",
      "            else:\n",
      ">               raise ValueError(f\"File format '{file_format}' is not supported.\")\n",
      "E               ValueError: File format 'txt' is not supported.\n",
      "\n",
      "../premier_league/s3_helpers.py:152: ValueError\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "mocker = <pytest_mock.plugin.MockerFixture object at 0x127e72a40>\n",
      "data =    a\n",
      "0  1\n",
      "1  2, file_path = 'path/to/file.txt', bucket = 'test-bucket'\n",
      "file_format = 'txt', profile_name = None, expected_call_format = None\n",
      "expected_exception = <class 'ValueError'>\n",
      "\n",
      "    @pytest.mark.parametrize(\n",
      "        \"data, file_path, bucket, file_format, profile_name, expected_call_format, expected_exception\",\n",
      "        [\n",
      "            # Test case for CSV format with custom profile\n",
      "            (pd.DataFrame({\"a\": [1, 2]}), \"path/to/file.csv\",\n",
      "             \"test-bucket\", \"csv\", \"custom-profile\", \"csv\", None),\n",
      "    \n",
      "            # Test case for Excel format with default profile\n",
      "            (pd.DataFrame({\"a\": [1, 2]}), \"path/to/file.xlsx\",\n",
      "             \"test-bucket\", \"excel\", None, \"excel\", None),\n",
      "    \n",
      "            # Test case for unsupported file format\n",
      "            (pd.DataFrame({\"a\": [1, 2]}), \"path/to/file.txt\",\n",
      "             \"test-bucket\", \"txt\", None, None, ValueError),\n",
      "        ],\n",
      "    )\n",
      "    def test_save_data_s3(mocker, data, file_path, bucket, file_format, profile_name, expected_call_format, expected_exception):\n",
      "        # Mock boto3 Session and s3 resource\n",
      "        mock_session = mocker.patch.object(boto3, \"Session\")\n",
      "        mock_s3_resource = mock_session.return_value.resource.return_value\n",
      "        mock_bucket = mock_s3_resource.Bucket.return_value\n",
      "    \n",
      "        if expected_exception:\n",
      "            with pytest.raises(expected_exception):\n",
      ">               s3_helpers.save_data_s3(data, file_path, bucket, file_format, profile_name)\n",
      "\n",
      "../tests/test_s3_helpers.py:83: \n",
      "_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ \n",
      "\n",
      "data =    a\n",
      "0  1\n",
      "1  2, file_path = 'path/to/file.txt', bucket = 'test-bucket'\n",
      "file_format = 'txt', profile_name = None\n",
      "\n",
      "    def save_data_s3(\n",
      "        data: pd.DataFrame,\n",
      "        file_path: str,\n",
      "        bucket: str = constants.S3_BUCKET,\n",
      "        file_format: str = \"csv\",\n",
      "        profile_name: Optional[str] = \"premier-league-app\",\n",
      "    ) -> None:\n",
      "        \"\"\"\n",
      "        Save a DataFrame to an S3 bucket.\n",
      "    \n",
      "        Parameters:\n",
      "        - data (pd.DataFrame): DataFrame to save.\n",
      "        - file_path (str): Path to the file within the S3 bucket.\n",
      "        - bucket (str): Name of the S3 bucket. Defaults to the value in\n",
      "        constants.S3_BUCKET.\n",
      "        - file_format (str): Format of the file to save ('csv', 'excel', etc.). Defaults\n",
      "        to \"csv\".\n",
      "        - profile_name (Optional[str]): AWS profile name. Defaults to \"premier-league-app\".\n",
      "    \n",
      "        Raises:\n",
      "        - NoCredentialsError: If AWS credentials are missing.\n",
      "        - PartialCredentialsError: If AWS credentials are incomplete.\n",
      "        \"\"\"\n",
      "        try:\n",
      "            if profile_name:\n",
      "                # Use the specified profile\n",
      "                session = boto3.Session(profile_name=profile_name)\n",
      "            else:\n",
      "                # Use environment variables or default profile\n",
      "                session = boto3.Session(\n",
      "                    aws_access_key_id=os.getenv(\"AWS_ACCESS_KEY_ID\"),\n",
      "                    aws_secret_access_key=os.getenv(\"AWS_SECRET_ACCESS_KEY\"),\n",
      "                    region_name=os.getenv(\"AWS_DEFAULT_REGION\"),\n",
      "                )\n",
      "            s3_resource = session.resource(\"s3\")\n",
      "            logger_config.logger.info(f\"Savng data to {bucket}/{file_path}\")\n",
      "            if file_format.lower() == \"csv\":\n",
      "                # Save DataFrame as a CSV\n",
      "                buffer = StringIO()\n",
      "                data.to_csv(buffer, index=False)\n",
      "                buffer.seek(0)\n",
      "                s3_resource.Bucket(bucket).put_object(Key=file_path, Body=buffer.getvalue())\n",
      "            elif file_format.lower() == \"excel\":\n",
      "                # Save DataFrame as an Excel file\n",
      "                buffer = BytesIO()\n",
      "                with pd.ExcelWriter(buffer, engine=\"xlsxwriter\") as writer:\n",
      "                    data.to_excel(writer, index=False)\n",
      "                buffer.seek(0)\n",
      "                s3_resource.Bucket(bucket).put_object(Key=file_path, Body=buffer.getvalue())\n",
      "            else:\n",
      "                raise ValueError(f\"File format '{file_format}' is not supported.\")\n",
      "        except NoCredentialsError as e:\n",
      "            logger_config.logger.error(\n",
      "                f\"An error occurred saving {bucket}/{file_path}: %s\", str(e)\n",
      "            )\n",
      "            raise NoCredentialsError(\n",
      "                \"\"\"Credentials not available. Make sure the profile name is\n",
      "                correct and the credentials are set up properly.\"\"\"\n",
      "            )\n",
      "        except PartialCredentialsError as e:\n",
      "            logger_config.logger.error(\n",
      "                f\"An error occurred saving {bucket}/{file_path}: %s\", str(e)\n",
      "            )\n",
      "            raise PartialCredentialsError(\n",
      "                \"Incomplete credentials. Please check your AWS configuration.\"\n",
      "            )\n",
      "        except Exception as e:\n",
      "            logger_config.logger.error(\n",
      "                f\"An error occurred saving {bucket}/{file_path}: %s\", str(e)\n",
      "            )\n",
      ">           raise Exception(f\"An unexpected error occurred: {str(e)}\")\n",
      "E           Exception: An unexpected error occurred: File format 'txt' is not supported.\n",
      "\n",
      "../premier_league/s3_helpers.py:172: Exception\n",
      "----------------------------- Captured stderr call -----------------------------\n",
      "2023-11-29 09:14:17,657 - PremierLeagueLogger - INFO - Savng data to test-bucket/path/to/file.txt\n",
      "2023-11-29 09:14:18,734 - PremierLeagueLogger - ERROR - An error occurred saving test-bucket/path/to/file.txt: File format 'txt' is not supported.\n",
      "------------------------------ Captured log call -------------------------------\n",
      "INFO     PremierLeagueLogger:s3_helpers.py:137 Savng data to test-bucket/path/to/file.txt\n",
      "ERROR    PremierLeagueLogger:s3_helpers.py:169 An error occurred saving test-bucket/path/to/file.txt: File format 'txt' is not supported.\n",
      "=========================== short test summary info ============================\n",
      "FAILED ../tests/test_s3_helpers.py::test_save_data_s3[data2-path/to/file.txt-test-bucket-txt-None-None-ValueError] - Exception: An unexpected error occurred: File format 'txt' is not supported.\n",
      "========================= 1 failed, 3 passed in 7.89s ==========================\n",
      "An error occurred while running tests: Tests failed. Please check.\n"
     ]
    }
   ],
   "source": [
    "# Run unit tests\n",
    "if constants.LOCAL_MODE:\n",
    "    path_prefix = '/Users/david@inawisdom.com/Documents/Training/premier_league'\n",
    "    test_path = (f'{path_prefix}/tests/run_tests.py')\n",
    "    r = code_quality_functions.run_unit_tests(\n",
    "        '../tests/test_s3_helpers.py',\n",
    "        'premier_league'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "25b4eca2-33a0-4fb6-907d-7177fbc17a79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "============================= test session starts ==============================\n",
      "platform darwin -- Python 3.10.1, pytest-7.4.3, pluggy-1.3.0 -- /Users/david@inawisdom.com/Documents/Training/premier_league/env_premier_league/bin/python\n",
      "metadata: {'Python': '3.10.1', 'Platform': 'macOS-14.1-arm64-arm-64bit', 'Packages': {'pytest': '7.4.3', 'pluggy': '1.3.0'}, 'Plugins': {'html': '2.0.1', 'rerunfailures': '12.0', 'xdist': '3.4.0', 'cov': '4.1.0', 'seleniumbase': '4.21.4', 'metadata': '3.0.0', 'mock': '3.12.0', 'anyio': '3.7.1', 'ordering': '0.6'}}\n",
      "rootdir: /Users/david@inawisdom.com/Documents/Training/premier_league\n",
      "plugins: html-2.0.1, rerunfailures-12.0, xdist-3.4.0, cov-4.1.0, seleniumbase-4.21.4, metadata-3.0.0, mock-3.12.0, anyio-3.7.1, ordering-0.6\n",
      "collecting ... collected 60 items / 1 error\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Module premier_league was previously imported, but not measured (module-not-measured)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "==================================== ERRORS ====================================\n",
      "______________ ERROR collecting tests/test_inference_pipeline.py _______________\n",
      "ImportError while importing test module '/Users/david@inawisdom.com/Documents/Training/premier_league/tests/test_inference_pipeline.py'.\n",
      "Hint: make sure your test modules/packages have valid Python names.\n",
      "Traceback:\n",
      "../../../../.pyenv/versions/3.10.1/lib/python3.10/importlib/__init__.py:126: in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "../tests/test_inference_pipeline.py:5: in <module>\n",
      "    from premier_league import inference_pipeline\n",
      "E   ImportError: cannot import name 'inference_pipeline' from 'premier_league' (/Users/david@inawisdom.com/Documents/Training/premier_league/notebooks/../premier_league/__init__.py)\n",
      "\n",
      "---------- coverage: platform darwin, python 3.10.1-final-0 ----------\n",
      "Name                                                                                                    Stmts   Miss  Cover   Missing\n",
      "-------------------------------------------------------------------------------------------------------------------------------------\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/__init__.py                     0      0   100%\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/app.py                         78     78     0%   1-215\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/code_quality_functions.py      45     45     0%   1-104\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/config.py                      18      4    78%   7-9, 15\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/constants.py                   32     32     0%   1-67\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/data_drift_functions.py        57     42    26%   14-17, 38-39, 59-113, 128-153\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/data_extraction.py            155    140    10%   5-8, 29-36, 59-95, 110-123, 149-171, 194-195, 216-247, 273-300, 324-351\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/email_functions.py             52     41    21%   9-11, 37-40, 69-77, 100-110, 138-165, 189-205\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/evaluation.py                  25     20    20%   8-9, 27-59\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/expectations_helpers.py       170    145    15%   14-16, 31, 42-46, 60-65, 78-81, 94-97, 113-134, 148-180, 220-254, 276-322, 338-356, 369-384, 392-426, 445-480, 495-497\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/logger_config.py               45     37    18%   1-45, 59-61, 69-88\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/mlflow_functions.py            45     32    29%   11-13, 32-37, 50-51, 66-76, 90-104, 118-119\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/postgres.py                    64     53    17%   10-12, 28-54, 81-112, 132-144, 163-174\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/prediction.py                  48     38    21%   10-11, 34-58, 72-81, 95-101, 123-132, 146-151\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/preprocessing.py               46     33    28%   7-9, 25-29, 46-50, 67-75, 84, 97-100, 116-119, 135-139\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/preprocessing_helpers.py      340    298    12%   22-42, 57-75, 91-116, 130-135, 153-158, 178-200, 218-235, 253-264, 281-302, 319-324, 340-347, 364-368, 386-390, 406-410, 426-430, 456-460, 477-480, 497-513, 526-529, 542-545, 558-561, 574-577, 593-602, 615-617, 631-633, 649-712, 728-743, 757-763, 779-786, 802-806, 822-828, 843-849, 868-882, 898-903, 917-918, 933-952, 968-992\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/s3_helpers.py                 172    172     0%   1-490\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/training.py                    97     97     0%   2-253\n",
      "/Users/david@inawisdom.com/Documents/Training/premier_league/premier_league/visualisations.py             246    246     0%   1-805\n",
      "-------------------------------------------------------------------------------------------------------------------------------------\n",
      "TOTAL                                                                                                    1735   1553    10%\n",
      "\n",
      "=========================== short test summary info ============================\n",
      "ERROR ../tests/test_inference_pipeline.py\n",
      "!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!\n",
      "========================= 1 warning, 1 error in 7.81s ==========================\n",
      "An error occurred while running tests: Tests failed. Please check.\n"
     ]
    }
   ],
   "source": [
    "if constants.LOCAL_MODE:\n",
    "    path_prefix = '/Users/david@inawisdom.com/Documents/Training/premier_league'\n",
    "    test_path = (f'{path_prefix}/tests/run_tests.py')\n",
    "    r = code_quality_functions.run_unit_tests(\n",
    "        '../tests/',\n",
    "        'premier_league',\n",
    "        'premier_league'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ced1791e-f519-42e7-bfc5-d6b85f8a28e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function run_unit_tests in module premier_league.code_quality_functions:\n",
      "\n",
      "run_unit_tests(test_file: str, src_folder: str, cov_folder: str = None, exclude_files: List[str] = []) -> Optional[int]\n",
      "    Run unit tests using pytest.\n",
      "    \n",
      "    Parameters:\n",
      "    - test_file (str): The path to the test file to run.\n",
      "    - src_folder (str): The path to the source code folder.\n",
      "    - cov_folder (str, optional): The folder to generate coverage for. Defaults to None.\n",
      "    - exclude_files (List[str], optional): List of file paths to exclude.\n",
      "    Defaults to an empty list.\n",
      "    \n",
      "    Returns:\n",
      "    - int or None: Returns the return code of pytest, or None if an error occurred.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(code_quality_functions.run_unit_tests)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eed5fe90-f73b-4742-8de9-108f1cfe5561",
   "metadata": {},
   "source": [
    "## Black and Flake8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d40a6805-60c4-4612-8e9d-d5354369596a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Black Output:\n",
      "Skipping .ipynb files as Jupyter dependencies are not installed.\n",
      "You can fix this by running ``pip install \"black[jupyter]\"``\n",
      "All done! ✨ 🍰 ✨\n",
      "40 files left unchanged.\n",
      "\n",
      "\n",
      "Flake8 Output:\n",
      "Done. Flake8 has no improvements!\n"
     ]
    }
   ],
   "source": [
    "if constants.LOCAL_MODE:\n",
    "    code_quality_functions.run_code_quality_checks(\n",
    "        '../premier_league',\n",
    "        exclude_folder='../premier_league/.ipynb_checkpoints/'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbfc5def-9d0d-422a-9b5f-9eb719ccdd26",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env_premier_league",
   "language": "python",
   "name": "env_premier_league"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
